tests/test_restriction.cpp

```diff
@@ -0,0 +1,288 @@
+                                double indx =
+                                    (2 * i + y < p_level.ntheta_int - 1) ? p_level.theta[2 * i + y + 1] : 2 * PI;
+
+                                k_q[k] = {indx - p_level.theta[2 * i + y],
+                                          p_level.theta[2 * i + y] - p_level.theta[2 * i + y - 1]};
+                            }
+                            else {
+                                switch (
+                                    y) { //based on the value of y and periodic boundary conditions we get different values for {k_q,k_{q-1}}
+                                case 1:
+                                    k_q[k] = {p_level.theta[2] - p_level.theta[1], p_level.theta[1] - p_level.theta[0]};
+                                case 0:
+                                    k_q[k] = {p_level.theta[1] - p_level.theta[0],
+                                              2 * PI - p_level.theta[p_level.ntheta_int - 1]};
+                                default:
+                                    k_q[k] = {2 * PI - p_level.theta[p_level.ntheta_int - 1],
+                                              p_level.theta[p_level.ntheta_int - 1] -
+                                                  p_level.theta[p_level.ntheta_int - 2]};
+                                }
+                            }
+                        }
+                        k += 1;
+                    }
+                }
+            }
+            /*values given by the operator. We multiply this with the grid-function value of the corresponding adjacent node*/
+
+            std::vector<double> vals{
+                std::get<1>(h_p[0]) * std::get<1>(k_q[0]) /
+                    ((std::get<0>(h_p[0]) + std::get<1>(h_p[0])) * (std::get<0>(k_q[0]) + std::get<1>(k_q[0]))),
+                std::get<1>(h_p[1]) / (std::get<0>(h_p[1]) + std::get<1>(h_p[1])),
+                std::get<1>(h_p[2]) * std::get<0>(k_q[2]) /
+                    ((std::get<0>(h_p[2]) + std::get<1>(h_p[2])) * (std::get<0>(k_q[2]) + std::get<1>(k_q[2]))),
+                std::get<1>(k_q[3]) / (std::get<0>(k_q[3]) + std::get<1>(k_q[3])),
+                std::get<0>(k_q[4]) / (std::get<0>(k_q[4]) + std::get<1>(k_q[4])),
+                std::get<0>(h_p[5]) * std::get<1>(k_q[5]) /
+                    ((std::get<0>(h_p[5]) + std::get<1>(h_p[5])) * (std::get<0>(k_q[5]) + std::get<1>(k_q[5]))),
+                std::get<0>(h_p[6]) / (std::get<0>(h_p[6]) + std::get<1>(h_p[6])),
+                std::get<0>(h_p[7]) * std::get<0>(k_q[7]) /
+                    ((std::get<0>(h_p[7]) + std::get<1>(h_p[7])) * (std::get<0>(k_q[7]) + std::get<1>(k_q[7])))};
+
+            double finval = u_test[(2 * j) * p_level.ntheta_int + (2 * i)];
+            for (int z = 0; z < 8; z++) {
+                finval += vals[z] * adjacent[z]; //accumulate all values in the coarse node
+            }
+            EXPECT_NEAR(finval, sol[j * ctheta_int + i], 1e-6)
```

Comment:
```suggestion
            EXPECT_NEAR(finval, sol[j * ctheta_int + i], 1e-12)
```
Can we make this succeed with 1e-12 too?